pkg/server/peer.go

```diff
@@ -282,10 +282,10 @@ func (peer *Peer) llgrFamilies() ([]bgp.RouteFamily, []bgp.RouteFamily) {
 
 func (peer *Peer) isLLGREnabledFamily(family bgp.RouteFamily) bool {
 	peer.fsm.lock.RLock()
-	defer peer.fsm.lock.RUnlock()
 	if !peer.fsm.pConf.GracefulRestart.Config.LongLivedEnabled {
 		return false
 	}
+	peer.fsm.lock.RUnlock()
```

Comment:
probably it’s best to split the function and have `llgrFamilies` called as a start of a new function  and put the defer back. That way you don't end up keeping the lock indefinitely should you return at line 286. Have you thought of that?

Reply:
I guess splitting the function means you’ have to acquire the lock twice and if the entire code is mean to be atomic, then probably there's room for a more elaborated fix.

Reply:
Good catch. Yes, a return on 286 would be bad. Let me go split the function and we'll see what it looks like.